### DISCUSSION TITLE
duda ejecución parte 3

### DISCUSSION THREAD

**Discussion by llambias** [IIC2133-PUC/2024-1#193] [Tarea 3] [answered]
Hola, al correr el test 3, simplemente correr int vecinos[NODOS_GRAFO]; me genera un segmentation error, debería cambiar algo?, hay que hacer un alloc o malloc? o mi estrategia no debería incluir una lista de ese tamaño simplemente?

**Comment by benjavicente** [comment]
Hola! Si pasa eso, te recomiendo cambiarlo por un malloc, porque puede que no le de el stack para guardar todos los nodos:

```diff
- int vecinos[NODOS_GRAFO];
+ int *vecinos = malloc(NODOS_GRAFO * siseof(int))  # revisa que esté ok, no lo probé
```

Si te sigue pasando un error, manda el detalle de este que se puede ver con valgrind.

**Reply by llambias** [reply]
hola me tira este error al compilar
error: expected expression before ‘int’
   74 |   int *vecinos = malloc(NODOS_GRAFO * siseof(int))
vencinos los estaba definiendo en esa lines y buscaba crear un array simplemente

**Reply by benjavicente** [reply]
No es la línea literal que mandé, por eso añadí el comentario.
```c
// Repito, REVISA que este ok.
// Lo importante es cambiar algo del stack (ie algo[largo]) a algo alocado en el heap (con malloc o similar)
int *vecinos = (int *)malloc(NODOS_GRAFO * sizeof(int)); 
```

**Reply by llambias** [reply]
eso soluciono el problema, lo que si  tengo este error que no estoy entendiendo como solucionarlo
`while(lista[revisar]!=revisar && lista[revisar] ==-2)`
 ```
Invalid read of size 4
==1057==    at 0x10973C: main (main.c:121)
==1057==  Address 0x4aa98f8 is 8 bytes before a block of size 40 alloc'd
==1057==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==1057==    by 0x1094BA: main (main.c:66)
```
antes de implementar lo que me dijiste tenia 
```
Conditional jump or move depends on uninitialised value(s)
==1043==    at 0x109769: main (main.c:121)
```
los cambios los hice en el array lista en lugar de vecinos, este lo borre

**Reply by benjavicente** [reply, marked as answer]
Ambos errores dicen que se está leyendo un valor que no es parte de lo que tienes definido. El hint principal es el `before`: parece que `revisar` puede tomar un valor `-1` o `-2`, lo que sería 8 bytes antes de lo que parte el arreglo (en C no existen índices negativos, sino que estarías leyendo memoria antes de donde está la lista).

**Reply by llambias** [reply]
eso ahora coloque un if para evitar indexar con -1 o-2 y soluciono el problema gracias
